ocaml/lib/snarky_js_bindings_lib.ml

```diff
@@ -294,6 +294,35 @@ module Snarky = struct
       | Unchecked s ->
           Poseidon_sponge.squeeze s |> Impl.Field.constant
   end
+
+  module Foreign_field = struct
+    module FF = Kimchi_gadgets.Foreign_field
+    module Range_check = Kimchi_gadgets.Range_check
+    module External_checks = FF.External_checks
+
+    type t = Impl.field FF.Element.Standard.t
+
+    type t_const = Impl.field FF.Element.Standard.limbs_type
+
+    (* high-level API of self-contained methods which do all necessary checks *)
+
+    let assert_valid_element (x : t) (p : t_const) : unit =
+      let external_checks = External_checks.create (module Impl) in
+      let _ = FF.valid_element (module Impl) external_checks x p in
+      FF.constrain_external_checks (module Impl) external_checks p
+
+    let add (x : t) (y : t) (p : t_const) : t =
```

Comment:
Oh, so by default you include all checks of additions? This means users of the API won't be able to chain additions efficiently, and will instead require 2 rows per FFAdd instead of 1 asymptotically.

Reply:
yes exactly!